src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_CAPACITY = 10;
+    private T[] array;
+    private int size = 0;
+
+    public ArrayList() {
+        array = (T[]) new Object[INITIAL_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == array.length) {
+            T[] newArray = (T[]) new Object[array.length + (array.length / 2)];
+            for (int i = 0; i < size; i++) {
+                newArray[i] = array[i];
+            }
+            array = newArray;
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+        if (size == array.length) {
+            T[] newArray = (T[]) new Object[array.length + (array.length / 2)];
+            for (int i = 0; i < index; i++) {
+                newArray[i] = array[i];
+            }
+            newArray[index] = value;
+            for (int i = index; i < size; i++) {
+                newArray[i + 1] = array[i];
+            }
+            array = newArray;
+        } else {
+            for (int i = size; i > index; i--) {
+                array[i] = array[i - 1];
+            }
+            array[index] = value;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() > array.length) {
+            T[] newArray = (T[]) new Object[size + list.size() + (array.length / 2)];
+            for (int i = 0; i < size; i++) {
+                newArray[i] = array[i];
+            }
+            array = newArray;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            array[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+        array[index] = value;
+    }
 
+    @Override
+    public T remove(T element) {
+        int index = indexOf(element);
+        if (index == -1) {
+            throw new NoSuchElementException("Element not found: " + element);
+        }
+        return remove(index);
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+
+        final T removedValue = array[index];
+        for (int i = index; i < size - 1; i++) {
+            array[i] = array[i + 1];
```

Comment:
Consider using `System.arraycopy()` instead of manually copying elements. This will improve performance and readability .